File: 0143-reorder-list/0143-reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode reverseList(ListNode head) {
        ListNode newHead = null;
        ListNode nextNode;

        while (head != null) {
            nextNode = head.next;
            head.next = newHead;
            newHead = head;
            head = nextNode;
        }

        return newHead;
    }

    public void reorderList(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode secondHalf = reverseList(slow.next);
        slow.next = null;

        ListNode firstHalf = head;
        while (secondHalf != null) {
            ListNode nextFirst = firstHalf.next;
            ListNode nextSecond = secondHalf.next;

            firstHalf.next = secondHalf;
            secondHalf.next = nextFirst;

            firstHalf = nextFirst;
            secondHalf = nextSecond;
        }
    }
}

File: 0143-reorder-list/0143-reorder-list.cpp
class Solution {
public:
    void reorderList(ListNode* head) {
       if (head->next == NULL || head->next->next == NULL) return; 
        
        stack<ListNode*> st;
        ListNode* temp = head;
        int size = 0;
        while (temp != NULL) {
            size++;
            st.push(temp);
            temp = temp->next;
        }
        
        temp = head;
        
        size = size / 2;
        while (size--) {
            ListNode *element = st.top();
            st.pop();
            
            element->next = temp->next;
            temp->next = element;
            temp = temp->next->next;
        }
        
        temp->next = NULL;
    }
};